compiler/rustc_ty_utils/src/layout.rs

```diff
@@ -146,6 +144,35 @@ fn univariant_uninterned<'tcx>(
     cx.calc.univariant(fields, repr, kind).map_err(|err| map_error(cx, ty, err))
 }
 
+fn validate_const_with_value<'tcx>(
+    _const: ty::Const<'tcx>,
```

Comment:
nit:

```suggestion
    const_: ty::Const<'tcx>,
```

so this doesn't accidentally hide the `dead_code` lint.

---

Also, can you make this function return the valtree directly? To clarify that this must extract a value or return an error.

Reply:
I wanted to force extraction but we have to cover two cases:
The first one:
```rust
validate_const_with_value(start, ty, cx)?
    .try_to_bits(tcx, cx.typing_env)
    .ok_or_else(|| error(cx, LayoutError::Unknown(ty)))?
```
The second one:
```rust
validate_const_with_value(count, ty, cx)?
                .to_valtree()
                .0
                .try_to_target_usize(tcx)
                .ok_or_else(|| error(cx, LayoutError::Unknown(ty)))?;
```

Note that the function `try_to_bits` cannot be called on `valtree` types, thus we cannot extract a `valtree` in `validate_const_with_value`. Am I missing something?

Reply:
Hm, right now we are matching for `ConstKind::Value` in this function here and then again in `try_to_bits` (which will call `try_to_valtree`). But I see now that changing this requires some non-trivial restructuring, so it's fine to leave this for now.

Reply:
Exactly, we can keep it as a follow-up to this PR.
One could trivially add a version of `try_to_bits` where we ensure that the `const` is actually a `valtree` avoiding to call the `try`, or something like that.